Backend/Gmfctn/Controllers/AchievementController.cs

```diff
@@ -1,75 +1,90 @@
+
+        public AchievementController(IAchievementService _AchievementService, IMapper _Mapper)
+        {
+            AchievementService = _AchievementService;
+            Mapper = _Mapper;
         }
         [HttpGet]
+        [Authorize(Roles = "User")]
         public async Task<ActionResult<IEnumerable<Achievement>>> GetAllAchievement(CancellationToken Cancel)
         {
             try
             {
-                var Achievements = await UnitOfWork.AchievementRepository.GetAll(Cancel);
+                var Achievements = await AchievementService.GetAllAchievements(Cancel);
                 if (Achievements == null)
+                {
                     return NotFound();
-                else
-                    return Ok(Achievements);
+                }
+                return Ok(Achievements);
             }
-            catch(Exception Exc) {
+            catch (TaskCanceledException Exc)
+            {
+                return Ok(Exc.Data);
+            }
+            catch (Exception Exc) {
                 throw Exc;
             }
             
         }
-        [HttpGet("{id}")]
-        public async Task<ActionResult<Achievement>> GetAchivementById(Guid Id, CancellationToken Cancel)
+        [HttpGet("{Id}")]
+        [Authorize(Roles = "User")]
+        public async Task<ActionResult<Achievement>> GetAchivementById( Guid Id, CancellationToken Cancel)
         {
             try {
-                var Achievement = await UnitOfWork.AchievementRepository.GetById(Id, Cancel);
+                var Achievement = await AchievementService.GetAchievementById(Id, Cancel);
                 if (Achievement != null)
                 {
                     return Ok(Achievement);
                 }
                 return NotFound();
             }
+            catch (TaskCanceledException Exc)
+            {
+                return Ok(Exc.Data);
+            }
             catch (Exception Exc)
             {
                 throw Exc;
             }
         }
         [HttpPost]
+        [Authorize(Roles = "Admin")]
         public async Task<ActionResult> CreateAchievement(AchievementCreateDTO Achievement, CancellationToken Cancel)
         {
             try
             {
-                if (!ModelsValidator.AchievementIsValid(Achievement))
-                    ModelState.AddModelError("achievement", "Contains errors in parametrs.");
-                if (ModelState.IsValid) 
+                if (await AchievementService.CreateAchievement(Achievement, Cancel))
                 {
-                    var _Achievement = _Mapper.Map<Achievement>(Achievement);
-                    _Achievement.Id = new Guid();
-                    await UnitOfWork.AchievementRepository.Create(_Achievement, Cancel);
-                    await UnitOfWork.SaveChangesAsync(Cancel);
                     return Ok();
                 }
-                return BadRequest(ModelState);
+                return BadRequest();
```

Comment:
I recommend using that part in the catch block. And In common, I see that your Services are return bool, but at my opinion, that is the weird solution. better let it return Task and handle some errors using try...catch blocks. So if operation passed good in try return OK() if no handle it in catch and there return BadRequest or whatever you need.

Reply:
Fixed